lib/Twig/Extension/Core.php

```diff
@@ -821,7 +821,27 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        while ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        }
+
+        if ($array instanceof Iterator) {
+            $keys = array();
+            $array->rewind();
+            while ($array->valid()) {
+                $keys[] = $array->key();
+                $array->next();
+            }
+
+            return $keys;
+        }
```

Comment:
It depends on the implementation of the Iterator. The case here is getting the keys only, we are not interested in the items.
For example consider something like this:
```php
<?php

final class Repository implements Iterator {

    private $dataBaseIds;
    private $position = 0;

    public function __construct(array $dataBaseIds) {

        $this->dataBaseIds = $dataBaseIds;
    }

    public function rewind() {
        $this->position = 0;
    }

    public function current() {
        return $this->loadItemFromDatabase($this->dataBaseIds[$this->position]);
    }

    public function key() {
        return $this->dataBaseIds[$this->position];
    }

    public function next() {
        ++$this->position;
    }

    public function valid() {
        return $this->position < count($this->dataBaseIds);
    }

    private function loadItemFromDatabase($id)
    {
        echo "\nloading ".$id;
        sleep(1);
        return 'item:'.$id;
    }
}

// Current Twig implementation
function getKeysCurrentTwigImplementation(Iterator $r)
{
    $keys = [];
    foreach ($r as $key => $item) {
        $keys[] = $key;
    }

    return $keys;
}

// PR'ed version
function getKeysPR(Iterator $r)
{
    $keys = [];

    $r->rewind();
    while ($r->valid()) {
        $keys[] = $r->key();
        $r->next();
    }

    return $keys;
}


$r = new Repository([8,9,10]);
$keys = getKeysCurrentTwigImplementation($r);
$keys = getKeysPR($r);

var_dump($keys);
```

The trick is that the `foreach` `key` `value` way always calls the `current` on the iterator, even when it is not used.
Ideal would be that PHP would allow for something like;
```php
foreach($iterator => $key;) {
}
```
so we would get the key, but not the value (another benefit would be that no 'unused variable' has to be made, keeping SCA's happy).